### PR TITLE
Implement GLV scalar multiplication algorithm

### PR DESCRIPTION
**Pull request by jules**

### CODE REVIEW

File: curves/src/templates/short_weierstrass_jacobian/projective.rs [hunk 1]
```diff
@@ -506,15 +506,116 @@ impl<P: Parameters> Mul<P::ScalarField> for Projective<P> {
...
+
+        let glv_endomorphism = |t: Affine<P>| -> Affine<P> {
+            if t.is_zero() {
+                return t;
+            }
+
+            let mut r = Affine::zero();
+            r.y = t.y;
+            r.x = phi(t.x);
+            r
```
Comment (Pratyush) [inline curves/src/templates/short_weierstrass_jacobian/projective.rs]:
```suggestion
        let glv_endomorphism = |mut t: Affine<P>| -> Affine<P> {
            t.x = phi(t.x);
            t
```

**Comment by jules** [comment]
Some preliminary bench results:

```
bls12_377: g1_mul_assign
                        time:   [87.016 µs 87.050 µs 87.091 µs]
                        change: [-47.347% -47.283% -47.229%] (p = 0.00 < 0.05)
                        Performance has improved.
bls12_377: g1_is_in_correct_subgroup
                        time:   [53.454 µs 53.485 µs 53.516 µs]
                        change: [-11.952% -11.651% -11.381%] (p = 0.00 < 0.05)
                        Performance has improved.
bls12_377: g2_mul_assign
                        time:   [268.61 µs 268.80 µs 269.00 µs]
                        change: [-56.031% -55.960% -55.888%] (p = 0.00 < 0.05)
                        Performance has improved.
```

File: curves/src/templates/short_weierstrass_jacobian/projective.rs [hunk 2]
```diff
@@ -499,22 +500,107 @@ impl<'a, P: Parameters> SubAssign<&'a Self> for Projective<P> {
...
+        const L: usize = 1 << (GLV_WINDOW_SIZE - 1);
+
+        let decomposition = other.decompose();
+
+        // Prepare tables.
+        let mut t_1 = Vec::with_capacity(L);
+        let double = Affine::from(self.double());
+        t_1.push(self);
+        for i in 1..L {
+            t_1.push(t_1[i - 1].add_mixed(&double));
```
Comment (ljedrz) [inline curves/src/templates/short_weierstrass_jacobian/projective.rs]: perhaps [iter::scan](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.scan) could be used here?

File: curves/src/templates/short_weierstrass_jacobian/projective.rs [hunk 3]
```diff
@@ -499,22 +500,107 @@ impl<'a, P: Parameters> SubAssign<&'a Self> for Projective<P> {
...
+                wnaf_1.iter_mut().for_each(|e| *e = -*e);
+            }
+            if !s2 {
+                wnaf_2.iter_mut().for_each(|e| *e = -*e);
+            }
+
+            (wnaf_1, wnaf_2)
+        };
+
+        let naf_add = |table: &Vec<Affine<P>>, naf: i32, acc: &mut Self| {
```
Comment (ljedrz) [inline curves/src/templates/short_weierstrass_jacobian/projective.rs]:
```suggestion
        let naf_add = |table: &[Affine<P>], naf: i32, acc: &mut Self| {
```

File: curves/src/templates/short_weierstrass_jacobian/projective.rs (hunk 2 above)
Comment (jules) [reply curves/src/templates/short_weierstrass_jacobian/projective.rs]: Possibly, though it might be a bit annoying to wrap all of the elements in `Option`s

Comment (ljedrz) [reply curves/src/templates/short_weierstrass_jacobian/projective.rs]: it's possible that the removal of bounds checks would make it faster, but only if `L` can be large

Comment (jules) [reply curves/src/templates/short_weierstrass_jacobian/projective.rs]: I see, I think L is fixed at 1 << 3 = 8. Is that large enough for it to have an impact?

Comment (ljedrz) [reply curves/src/templates/short_weierstrass_jacobian/projective.rs]: nope, most likely not :+1:

File: fields/src/fp_256.rs [hunk 4]
```diff
@@ -376,6 +376,86 @@ impl<P: Fp256Parameters> PrimeField for Fp256<P> {
...
     type Parameters = P;
 
+    #[inline]
+    fn decompose(&self) -> (Self, Self, bool, bool) {
+        const Q1: [u64; 4] = [9183663392111466540, 12968021215939883360, 3, 0];
+        const Q2: [u64; 4] = [13, 0, 0, 0];
+        let b1 = Self::from_repr(BigInteger([725501752471715840, 4981570305181876225, 0, 0])).unwrap();
+        let b2 = Self::from_repr(BigInteger([725501752471715841, 4981570305181876225, 0, 0])).unwrap();
+        let r128 = Self::from_repr(BigInteger([0xffffffffffffffff, 0xffffffffffffffff, 0, 0])).unwrap();
+        const HALF_R: [u64; 8] = [0, 0, 0, 0x8000000000000000, 0, 0, 0, 0];
```
Comment (Pratyush) [inline fields/src/fp_256.rs]: These constants are specific to BLS12-377, right? We should move these to `BLS12Parameters`, and pass these in to `decompose`

File: fields/src/fp_384.rs [hunk 5]
```diff
@@ -434,6 +434,11 @@ impl<P: Fp384Parameters> PrimeField for Fp384<P> {
     type BigInteger = BigInteger;
     type Parameters = P;
 
+    #[inline]
+    fn decompose(&self) -> (Self, Self, bool, bool) {
+        (Self::zero(), Self::zero(), false, false)
```
Comment (Pratyush) [inline fields/src/fp_384.rs]:
```suggestion
        unimplemented!()
```

File: curves/src/bls12_377/tests.rs [hunk 6]
```diff
@@ -616,6 +617,31 @@ fn test_fq12_mul_by_034() {
...
+    //     7829691328573998628,
+    //     14277141908032039045,
+    //     12498936762639071483,
+    //     4759044196164073388,
+    //     3898951160843530891,
+    //     117220051666006481,
+    // ]))
+    // .unwrap();
+    // println!("{:?}", acc.x);
+    // println!("{:?}", x);
```
Comment (Pratyush) [inline curves/src/bls12_377/tests.rs]: Is the commented out part testing on a special edge-case?

File: curves/src/bls12_377/g2.rs [hunk 7]
```diff
@@ -89,6 +104,13 @@ impl ShortWeierstrassParameters for Bls12_377G2Parameters {
     ) -> bool {
         p.mul_bits(BitIteratorBE::new(Self::ScalarField::characteristic())).is_zero()
     }
+
+    fn glv_endomorphism(
```
Comment (Pratyush) [inline curves/src/bls12_377/g2.rs:127]: Something that we might want to follow up on is that for G2 there exists an even more efficient decomposition relying on the GLS endomorphism (https://eprint.iacr.org/2008/194)

File: curves/src/bls12_377/tests.rs (hunk 6 above)
Comment (jules) [reply curves/src/bls12_377/tests.rs]: Oh yeah, this was part of some debugging efforts. I'll take it out :)

File: fields/src/fp_256.rs (hunk 4 above)
Comment (jules) [reply fields/src/fp_256.rs]: I'll put them on the `ShortWeierstrassParameters` as it's not possible to access `BLS12Parameters` from `Fp256` or `Projective` as far as I can tell.

File: curves/src/bls12_377/g2.rs (hunk 7 above)
Comment (jules) [reply curves/src/bls12_377/g2.rs:127]: Would you know the speedup this would give? If it's big I would consider doing it for this PR, otherwise might just want to create an issue.

Comment (Pratyush) [reply curves/src/bls12_377/g2.rs:127]: Let's create an issue, since it's not really on the critical path for anything (there's only one G2 mul in Marlin verification)

Comment (jules) [reply curves/src/bls12_377/g2.rs:127]: Done!

File: fields/src/fp_256.rs (hunk 4 above)
Comment (Pratyush) [reply fields/src/fp_256.rs]: Hm I would rather put a `mul` method in `ShortWeierstrassParameters`, so that we can specify it when instantiating `Bls12377G1Parameters`. See for example https://github.com/arkworks-rs/algebra/blob/402e7f9603fca7a68b86baf296b6feaf904939f5/ec/src/models/short_weierstrass/mod.rs#L82

Comment (jules) [reply fields/src/fp_256.rs]: Hmm, so as in a `mul_by_b1`, `mul_by_q1` etc. sort of thing?

Comment (Pratyush) [reply fields/src/fp_256.rs]: I mean by a `mul_projective(a: Projective<Self>, b: Self::ScalarField) → Projective<Self>`. This way you can encapsulate the GLV portion in the curve-specific parameters

**Comment by howardwu** [comment]
^Rebased here.

**Review by Pratyush — APPROVED** [review]
This LGTM! The last (optional) change that might be good would be to use a helper method to share code between `G1Parameters::mul_projective` and `G2Parameters::mul_projective`. However this can be done in a second PR.
